frame/identity/src/lib.rs

```diff
@@ -473,40 +479,65 @@ decl_module! {
 		/// - `O(R)` where `R` registrar-count (governance-bounded).
 		/// - One storage mutation (codec `O(R)`).
 		/// - One event.
+		/// - Benchmark: 78.71 + R * 0.965 µs (min square analysis)
 		/// # </weight>
-		#[weight = SimpleDispatchInfo::FixedNormal(MINIMUM_WEIGHT)]
-		fn add_registrar(origin, account: T::AccountId) {
+		#[weight = FunctionOf(
+			|(_, &old_count): (&T::AccountId, &u32)| {
+				T::DbWeight::get().reads_writes(1, 1)
+				+ 78_710_000 // constant
+				+ 965_000 * (old_count as Weight) // R
+			},
+			DispatchClass::Normal,
+			true
+		)]
+		fn add_registrar(origin, account: T::AccountId, old_registrar_count: u32) -> DispatchResult {
 			T::RegistrarOrigin::try_origin(origin)
 				.map(|_| ())
 				.or_else(ensure_root)?;
 
-			let i = <Registrars<T>>::mutate(|r| {
-				r.push(Some(RegistrarInfo { account, fee: Zero::zero(), fields: Default::default() }));
-				(r.len() - 1) as RegistrarIndex
-			});
+			let i = <Registrars<T>>::try_mutate(|registrars| -> Result<RegistrarIndex, DispatchError> {
+				ensure!(registrars.len() as u32 <= old_registrar_count, "invalid count");
+				ensure!((registrars.len() as u32) < T::MaxRegistrars::get(), Error::<T>::TooManyRegistrars);
+				registrars.push(Some(RegistrarInfo { account, fee: Zero::zero(), fields: Default::default() }));
+				Ok((registrars.len() - 1) as RegistrarIndex)
+			})?;
 
 			Self::deposit_event(RawEvent::RegistrarAdded(i));
+
+			Ok(())
 		}
 
 		/// Set an account's identity information and reserve the appropriate deposit.
 		///
 		/// If the account already has identity information, the deposit is taken as part payment
 		/// for the new deposit.
 		///
-		/// The dispatch origin for this call must be _Signed_ and the sender must have a registered
-		/// identity.
+		/// The dispatch origin for this call must be _Signed_.
 		///
 		/// - `info`: The identity information.
 		///
 		/// Emits `IdentitySet` if successful.
 		///
 		/// # <weight>
-		/// - `O(X + X' + R)` where `X` additional-field-count (deposit-bounded and code-bounded).
+		/// - `O(X + X' + J)`
+		///   - where `X` additional-field-count (deposit-bounded and code-bounded)
+		///   - where `J` judgements-count (registrar-count-bounded)
 		/// - At most two balance operations.
-		/// - One storage mutation (codec-read `O(X' + R)`, codec-write `O(X + R)`).
+		/// - One storage mutation (codec-read `O(X' + J)`, codec-write `O(X + J)`).
 		/// - One event.
+		/// - Benchmark: 136.6 + J * 0.62 + X * 2.62 µs (min square analysis)
 		/// # </weight>
-		#[weight = SimpleDispatchInfo::FixedNormal(50_000_000)]
+		#[weight = FunctionOf(
+			|(info,): (&IdentityInfo,)| {
+				T::DbWeight::get().reads_writes(1, 1)
+				+ T::DbWeight::get().reads_writes(1, 1) // balance ops
+				+ 136_600_000 // constant
+				+ 2_622_000 * (info.additional.len() as Weight) // X
+				+ 624_000 * (T::MaxRegistrars::get() as Weight) // J
+			},
+			DispatchClass::Normal,
+			true
+		)]
```

Comment:
If this is `Normal` and `true`, you dont  need to use `FucntionOf`. You can just write the weight formula directly. 👍